Initializing Audit Log Drawer

robertandremitchell opened the pull request:
# PULL REQUEST

## Summary

Adds a drawer on click of individual audit row.

It also starts to build out infrastructure for rendering names instead of usernames for the name and also format our functions into readable, dynamic strings for the action column.

<img width="1252" alt="Screenshot 2025-04-11 at 11 17 15 AM" src="https://github.com/user-attachments/assets/fddbb573-9678-4ea5-8b79-21caa79a04f0" />
<img width="729" alt="Screenshot 2025-04-11 at 11 17 22 AM" src="https://github.com/user-attachments/assets/9e5a3b7b-c7d9-49ca-bc6c-52519e250918" />


## Related Issue

Fixes #434 

## Additional Information

Anything else the review team should know?

## Checklist

- [ ] Descriptive Pull Request title
- [ ] Link to relevant issues
- [ ] Provide necessary context for design reviewers
- [ ] Ensure test coverage is above agreed upon threshold
- [ ] Update documentation

mikang commented:
Thanks, @robertandremitchell !! Here are some comments:
- Could you please remove the pagination component at the bottom if there are no results or only 1 page? It should appear only when there are 2 pages or more. 
![image](https://github.com/user-attachments/assets/1148a19c-d660-4d44-a877-047e79dc40bc)
- When hovering over a row in the audit log, 1) there should be a blue highlight and 2) the cursor should turn into a pointing hand to indicate that the row is clickable
- There should be no space between the "Full JSON request" and QC Admin, etc.

Regarding the date dropdown:
- "Clear" should be at the bottom of the dropdown left aligned
- There should be 16px of space between the two date pickers 

Currently it looks like this
<img width="260" alt="image" src="https://github.com/user-attachments/assets/b31ad36b-6ccf-4c74-900d-77fb8f52e3dd" />

robertandremitchell commented:
<img width="1302" alt="Screenshot 2025-04-14 at 1 44 32 PM" src="https://github.com/user-attachments/assets/0e0a4ae0-212c-422b-bdd4-25d0874c0d63" />

nickclyde reviewed (commented):
Amazing work, Rob! Ran it locally, really looks great. Mostly working as expected, though ran into two small bugs:

1. If I choose 50 per page, it works fine, but if I first go to the 2nd or 3rd page and then choose 50 per page, I end up with a blank page, all the table entries disappear
2. It's showing "Invalid date" for all of my entries - not sure if this is a me issue though? Maybe I need to clear my table?

robertandremitchell commented:
> Amazing work, Rob! Ran it locally, really looks great. Mostly working as expected, though ran into two small bugs:
> 
> 1. If I choose 50 per page, it works fine, but if I first go to the 2nd or 3rd page and then choose 50 per page, I end up with a blank page, all the table entries disappear
> 2. It's showing "Invalid date" for all of my entries - not sure if this is a me issue though? Maybe I need to clear my table?

ack, let me look at #1. I'm guessing it's being overzealous in the new code I added to hide pagination/actions per page.

re: #2, can you share a screenshot? The start/end dates are rather permissive on formats (MM/DD/YYYY, YYYY-MM-DD, etc.) so curious to see what's happening there.

nickclyde commented:
![Screenshot 2025-04-15 at 1 10 22 PM](https://github.com/user-attachments/assets/23b1c9a8-8fea-4bf1-8482-0b65d8a60d80)

Sure thing, here's the screenshot, and here's what my db looks like:

<img width="1573" alt="Screenshot 2025-04-15 at 1 11 05 PM" src="https://github.com/user-attachments/assets/d708e2b7-9397-42bf-8dae-924d57f4708a" />

robertandremitchell commented:
> ![Screenshot 2025-04-15 at 1 10 22 PM](https://private-user-images.githubusercontent.com/9121162/434019980-23b1c9a8-8fea-4bf1-8482-0b65d8a60d80.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3NDk4NjAsIm5iZiI6MTc0NDc0OTU2MCwicGF0aCI6Ii85MTIxMTYyLzQzNDAxOTk4MC0yM2IxYzlhOC04ZmVhLTRiZjEtODQ4Mi0wYjY1ZDhhNjBkODAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTVUMjAzOTIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzUyMjViZDZiNGExZGI0YzgwZjI2NThlZDA1N2JjN2JkMzdmOTExYzc5ZTgyMmE3MjI0ZGViMzEzMzc2ZGMyMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.bt5QZD6rwaY28ZcBMK2pUQMAkVWN9zPnBZJp6Ing1Xw)
> 
> Sure thing, here's the screenshot, and here's what my db looks like:
> 
> <img alt="Screenshot 2025-04-15 at 1 11 05 PM" width="1573" src="https://private-user-images.githubusercontent.com/9121162/434020233-d708e2b7-9397-42bf-8dae-924d57f4708a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3NDk4NjAsIm5iZiI6MTc0NDc0OTU2MCwicGF0aCI6Ii85MTIxMTYyLzQzNDAyMDIzMy1kNzA4ZTJiNy05Mzk3LTQyYmYtOGRhZS05MjRkNTdmNDcwOGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTVUMjAzOTIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGEzMDAyMGZjYjg1NjEzZmM3NjEwNjBlMTM2NDQ3YjdlNDdkNzBkMDQ5NDA3ZWJmMzdhNWIwZjM2NzFlMzM1OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zSB1cldBUB_2xOuwki08wczr3OJcuaJ9RAq2Os_3QWY">

interesting...having trouble replicating locally but can keep looking. My initial thought is that it has to do with the code I added here:

```typescript
  useEffect(() => {
    async function fetchAuditLogs() {
      const logs = await getAuditLogs();
      return logs.map((log) => ({
        ...log,
        createdAt: new Date(log.createdAt + "Z"),
      }));
    }
```

@nickclyde , would you be able to test the code with this?

```typescript
  useEffect(() => {
    async function fetchAuditLogs() {
      const logs = await getAuditLogs();
      return logs;
    }
```

The reason I added the `Z` is that I found that when formatting the date below that it defaults to UTC offset 0 (GMT) because it interprets the Z-less date that way, while the Z-ful date will look at local system time and display the datetime in the appropriate timezone.

However, I wonder if this is not an OS-agnostic solution to this issue. If the original code works, I can try to find an alternative method for displaying datetimes in the user's timezone.

robertandremitchell commented:
@nickclyde , I pushed a more comprehensive rewrite of how to try to handle datetime stamps. lmk if you still run into `Invalid Date`; I'm hoping this should do the trick.

katyasoup commented on query-connector/src/app/(pages)/auditLogs/components/auditLogMaps.ts:
```diff
@@ -0,0 +1,142 @@
+/**
+ * Maps action types to their labels and formats.
+ */
+export const auditLogActionTypeMap: Record<string, auditLogActionTypeMapping> =
```
nice!

fzhao99 replied:
+1!

Consideration for future improvement, no action needed: I'm a little nervous as we build out more of this feature that there's so much custom logic for transforming data values at the UI level, and depending on churn of this code, worried about DRYness. Think it might be overoptimizing to try and generalize things from here, but if we find ourselves updating this constant too much, would love to look into generalizing the formatting logic here (maybe by putting some stricter regs around the method names so that they're more consistently verbDescriptorObject?)

robertandremitchell replied:
ticket here: https://linear.app/skylight-cdc/issue/QUE-246/spike-improvement-ideas-for-audit-log-mapping

katyasoup reviewed (approved):
looks great! i'm not seeing any issues with the date formatting on my end

fzhao99 commented on query-connector/src/app/(pages)/auditLogs/page.tsx:
```diff
@@ -29,13 +36,26 @@ const AuditLogs: React.FC = () => {
+      return logs.map((log) => {
+        let str = String(log.createdAt).trim();
+
+        if (!str.includes("T")) str = str.replace(" ", "T");
```
non-blocking suggestion: Could we add a comment here for why this is necessary? Guessing it's a timezone / stamp conversion thing, but want to be kind to future devs :)

robertandremitchell replied:
makes sense, added! and yeah, it's for timezone display

fzhao99 commented on query-connector/src/app/(pages)/auditLogs/page.tsx:
```diff
@@ -47,11 +67,23 @@ const AuditLogs: React.FC = () => {
 
   const uniqueNames = useMemo(
-    () => Array.from(new Set(logs.map((log) => log.author))).sort(),
+    () =>
```
note for future, no action needed: guessing these may become opportunities for performance tuning / better SQL as the audit table gets big, since the set operation in JS is pretty expensive. We should keep an eye on this!

robertandremitchell replied:
added backlog ticket: https://linear.app/skylight-cdc/issue/QUE-245/backlog-investigation-write-less-expensive-way-to-do-sortset-for-audit

fzhao99 reviewed (approved):
lgtm! great work on this rob :)

nickclyde reviewed (approved):
Great work Rob, thanks for fixing the date issue!
